contracts/PBAB+Collabs/GenArt721CoreV2_PBAB_FLEX.sol

```diff
@@ -0,0 +1,893 @@
+     *  0 - IPFS
+     *  1 - ARWEAVE
+     */
+    function updateProjectExternalAssetDependency(
+        uint256 _projectId,
+        uint256 _index,
+        string calldata _cid,
+        ExternalAssetDependencyType _dependencyType
+    )
+        external
+        onlyUnlockedProjectExternalAssetDependencies(_projectId)
+        onlyArtistOrWhitelisted(_projectId)
+    {
+        require(
+            _index < projects[_projectId].externalAssetDependencyCount,
+            "Asset index out of range"
+        );
+        projects[_projectId].externalAssetDependencies[_index].cid = _cid;
+        projects[_projectId]
+            .externalAssetDependencies[_index]
+            .dependencyType = _dependencyType;
+        emit ExternalAssetDependencyUpdated(
+            _projectId,
+            _index,
+            _cid,
+            _dependencyType
+        );
+    }
+
+    /**
+     * @notice Removes external asset dependency for project `_projectId` at index `_index`.
+     * Assets with indices higher than `_index` can have their indices adjusted as a result of this operation.
+     * @param _projectId Project to be updated.
+     * @param _index Asset index
+     */
+    function removeProjectExternalAssetDependency(
+        uint256 _projectId,
+        uint256 _index
+    )
+        external
+        onlyUnlockedProjectExternalAssetDependencies(_projectId)
+        onlyArtistOrWhitelisted(_projectId)
+    {
+        require(
+            _index < projects[_projectId].externalAssetDependencyCount,
+            "Asset index out of range"
+        );
+
+        uint256 lastElementIndex = projects[_projectId]
+            .externalAssetDependencyCount - 1;
+
+        projects[_projectId].externalAssetDependencies[_index] = projects[
+            _projectId
+        ].externalAssetDependencies[lastElementIndex];
+        delete projects[_projectId].externalAssetDependencies[lastElementIndex];
+
+        projects[_projectId].externalAssetDependencyCount = lastElementIndex;
+
+        emit ExternalAssetDependencyRemoved(_projectId, _index);
+    }
+
+    /**
+     * @notice Adds external asset dependency for project `_projectId`.
+     * @param _projectId Project to be updated.
+     * @param _cid Asset cid (Content identifier).
+     * @param _dependencyType Asset dependency type.
+     *  0 - IPFS
+     *  1 - ARWEAVE
+     */
+    function addProjectExternalAssetDependency(
+        uint256 _projectId,
+        string calldata _cid,
+        ExternalAssetDependencyType _dependencyType
+    )
+        external
+        onlyUnlockedProjectExternalAssetDependencies(_projectId)
+        onlyArtistOrWhitelisted(_projectId)
+    {
+        ExternalAssetDependency memory asset = ExternalAssetDependency({
+            cid: _cid,
+            dependencyType: _dependencyType
+        });
+        projects[_projectId].externalAssetDependencies[
+            projects[_projectId].externalAssetDependencyCount
```

Comment:
nit: loading `projects[_projectId].externalAssetDependencyCount++` into memory will save you gas by eliminating redundant SLOAD